Offcanvas as component

GeoSot opened the pull request:
I did a first approach on an offcanvas component.
Αll of us, have used custom solutions, but maybe an official version,  would be more stable.

Do you find it interesting? May I continue on this?

Preview: <https://deploy-preview-29017--twbs-bootstrap.netlify.app/docs/5.0/components/offcanvas/>

TODO:

- [x] we need to see if we can stop body scroll
- [x] maybe blur the bg if it doesn't have any performance hit
- [x] tweak blur, the current blur is too much IMO (XhmikosR)
- [x] Remove height from bottom offcanvas
- [x] Provide some default theming (white background, maybe a border or shadow like the modals)
- [x] Disable body scrolling by default.
- [x] measure FPS when blurring the bg (@XhmikosR  i need guidance)
- [x] update offcanvas example to use this (also move it away from the experimental category)
- [x] reindent and fix missing end tags in the doc examples
- [x] switch to `event.key`
- [x] add more content in right, bottom examples
- [x] `<div class="h4">` should be changed to `h4` or similar?
- [x] Make sure the element is hidden and not offscreen when inactive 
- [x] enforce focus, only on open element
- [x] Enable blur by default?
- [x] Revert the scrollbar decoupling
- [x] Consistent behavior of modals and offcanvas (eg. same way of setting a backdrop) #32439
- [x] [non-text elements inside button](https://github.com/twbs/bootstrap/pull/29017#issuecomment-706302016) related issue #32372, #31646
- [x] Make sure all data attributes and any CSS vars are prefixed
- [x] Bring back the backdrop option, but keep it enabled by default
- [x] Make sure we aren't using any recently changed mr, etc classes
- [ ] Should the backdrop transition be the same as the offcanvas one? Right now, backdrop shows instantly but finishes hiding after the offcanvas is closed .(GeoSot: it will be changed when #32439 is merged)
- [x] Decide about the name: `offcanvas`, `off-canvas` etc
- [x] Maybe rename the classes to consider RTL? Not sure if it's needed
- [x] Fix BrowserStack tests
- [x] spell checking from a native English speaker (@mdo?)
- [x] check accessibility (@patrickhlauke)
- [ ] **clean up/squash patches manually**

Fixes #27033, fixes #24718.

---

mdo's TODO:

- [x] Finish and merge this PR
  - [x] Undo the Offcanvas example change for now (we'll replace that later)
- [x] Close any other PRs
- [x] Iterate on the docs a bit more
- [x] Detach and de-dupe the backdrop in #32439

Johann-S commented on js/src/offcanvas.js:
```diff
@@ -0,0 +1,194 @@
+ */
+
+import { getSelectorFromElement, jQuery as $ } from './util/index'
+
```
trailing line

Johann-S commented on js/src/offcanvas.js:
```diff
@@ -0,0 +1,194 @@
+ * ------------------------------------------------------------------------
+ */
+
+class Offcanvas {
```
I would name it: `OffCanvas` 🤔 but maybe I'm wrong ^^

Johann-S commented on js/src/offcanvas.js:
```diff
@@ -0,0 +1,194 @@
+
+  // Static
+
+  static _getParentFromElement(element) {
```
No need for a static method just a private one is enough

Johann-S commented on js/src/offcanvas.js:
```diff
@@ -0,0 +1,194 @@
+
+    EventHandler.on(this._element, Event.CLICK_DISMISS, Selector.DATA_DISMISS, event => this.hide(event))
+    EventHandler.on(document, Event.CLICK_DATA_API, event => {
+      const target = Offcanvas._getParentFromElement(event.target)
```
I'm more in favor of adding a css class to the body element to detect when a OffCanvas is open, it'll be faster to check if the body contains this class, and if it's the case we can try to figure out the parent of the target.

GeoSot replied:
Please be comfortable, to make corrections, as you have better knowledge of the code functionalities

Johann-S commented on js/src/offcanvas.js:
```diff
@@ -0,0 +1,194 @@
+      if (!data) {
+        data = new Offcanvas(this)
+      }
+
```
it lacks of the check for undefined methods, it's better to allow jQuery folks to call other methods too
```js
    if (typeof config === 'string') {
      if (typeof data[config] === 'undefined') {
        throw new TypeError(`No method named "${config}"`)
      }

      data[config]()
    }

GeoSot replied:
I think it's ok now

Johann-S reviewed (changes requested):
Hi @GeoSot it's very interesting thanks! 👍 
We are totally interested in this plugin

BTW you'll have to add unit tests

mdo commented:
Awesome! Haven’t dived into the code yet m, but obviously there’s a huge need for this. Can we ensure we have support for left, right, and bottom slide-in? Can we also make sure to stop the body scroll behind the sidebar?

patrickhlauke commented:
a crucial aspect of a properly working off-canvas is that focus should move to the offcanvas, remain trapped inside it until it's closed, and ideally the underlying document should be hidden via `aria-hidden="true"`

patrickhlauke reviewed (changes requested):
keyboard and assistive technology behavior must be addressed before this can be considered. conceptually, the off-canvas behaves pretty much like a dialog.

GeoSot commented:
Hello Guys. 

I really love this framework, and I want to contribute.

As you can see,  I am a bit slower than the requested changes, and really I don't have any experience on unit testing.
So I would appreciate any help

Johann-S commented:
@GeoSot you should first focus on request changes and after I'll help you for the unit tests don't worry 😉

MartijnCuppens commented:
> I don't have any experience on unit testing

Just have a look at the existing tests (`js/tests/unit`) and try to figure out how it works (for me, that worked 😛)

Other concerns:
- I wouldn't nest the selectors (like `&-right`), we don't do that either for our modals. The full classname also makes searching for classes easier.
- I would provide some default theming (like a default background) and create a theming section on the page with the colored components.
- I spotted some inline styles like `style="z-index: 1072;"`, we should avoid that.
- Accessibility wise I would like to close the drawer when I press <kbd>ESC</kbd> key.

GeoSot commented:
>     * I spotted some inline styles like `style="z-index: 1072;"`, we should avoid that.
Only because the doc header has 1071 z-index

Johann-S commented:
Hi @GeoSot, do not hesitate to hit me up if you're stuck somewhere or if you did everything you planed to 😉

GeoSot commented:
Thank u @Johann-S 

You may have a look on  Js code and leave me some feedback. (Feel free to change things)

Johann-S commented on js/src/offcanvas.js:
```diff
@@ -0,0 +1,208 @@
+    this._element.setAttribute('tabindex', savedTabIndex || 1)
+  }
+
+  _handleClosing() {
```
I would call it: `_addEventListener`

GeoSot replied:
I didn't got the point...

Johann-S commented on js/src/offcanvas.js:
```diff
@@ -0,0 +1,208 @@
+ */
+
+EventHandler.on(document, Event.CLICK_DATA_API, Selector.DATA_TOGGLE, function (event) {
+  if (['A', 'AREA'].includes(this.tagName)) {
```
`includes` isn't supported by the browsers we target 😟 you can use `.indexOf` instead

Johann-S commented:
@GeoSot we switched from QUnit to Jasmine, so you'll have to update your branch and use Jasmine to write your unit tests, do not hesitate to take a look at other plugins

Johann-S commented:
Hi @GeoSot, I improved your unit tests, so the JavaScript part should be ok 👍 now we have to check the docs and maybe the SCSS ? @twbs/css-review

MartijnCuppens commented on scss/_offcanvas.scss:
```diff
@@ -0,0 +1,40 @@
+}
+
+.offcanvas.show {
+  transform: translate(0, 0);
```
This can be simplified:
```suggestion
  transform: none;
```

MartijnCuppens commented:
> I preferred to keep it simple and focus on functionality and minimalism. If it succeed to be merged we can add some styling

Our other components also work out of the box without additional utilities for backgrounds (like the modal). For consistency, I would also like to see this here.

We should also ditch the heights for the bottom variant. Also there's some visible outline when the component is focused which need to be removed, like we do with the modals.

This PR also needs an approval from @patrickhlauke, to make sure everything is accessible.

XhmikosR commented:
Before the final merge, I'd like to clean up/squash any patches manually.

XhmikosR commented on js/src/offcanvas/offcanvas.js:
```diff
@@ -0,0 +1,205 @@
+
+      EventHandler.trigger(this._element, Event.SHOWN, { relatedTarget })
+      document.body.classList.add(ClassName.OPEN)
+    }, getTransitionDurationFromElement(this._element) + 9)
```
What's this magic number `9`?

XhmikosR commented on js/src/offcanvas/offcanvas.js:
```diff
@@ -0,0 +1,205 @@
+    const savedTabIndex = this._element.getAttribute('tabindex')
+    this._element.setAttribute('tabindex', '0')
+    this._element.focus()
+    this._element.setAttribute('tabindex', savedTabIndex || 1)
```
Shouldn't this be `'1'`?

XhmikosR commented:
Also, we need to see if we can stop body scroll like @mdo said above https://github.com/twbs/bootstrap/pull/29017#issuecomment-510986442

And maybe blur the bg if it doesn't have any performance hit.

GeoSot commented:
> Our other components also work out of the box without additional utilities for backgrounds (like the modal). For consistency, I would also like to see this here.

Maybe like the [navBar](https://getbootstrap.com/docs/4.3/components/navbar/#color-schemes)?  Honestly, I don't mind, if anybody wants to commit some styling changes, but anyway I cannot make this choice by myself


>What's this magic number 9?

Was the ugly way to keep the 'offcanvas-open' class, on body while toggling two different elements. I 've made some changes (maybe will need some refactoring)

XhmikosR commented:
Add a TODO list in the first post.

GeoSot commented:
@XhmikosR is offcanvas example needed anymore? 
With offcanvas as a basic component, maybe the example is redundant

XhmikosR commented:
You have a point there @GeoSot, but it's nice to have it as an individual example like we have navs etc.

XhmikosR commented:
Alright, I rebased this and cleaned up the patches.

Only a few things to do, and then another round of reviews.

XhmikosR commented:
The examples need to be reindended BTW.

XhmikosR commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,200 @@
+---
+layout: docs
+title: Offcanvas
+description: Toggle the visibility sidebars across your project with a few classes and our JavaScript plugins.
```
This needs some rewording. Also `plugins` -> `plugin`.

MartijnCuppens commented on scss/_offcanvas.scss:
```diff
@@ -0,0 +1,69 @@
+  top: auto;
+  right: 0;
+  width: 100%;
+  height: $offcanvas-vertical-height-mobile;
```
I would remove the height, it makes more sense if the component scales to the content

GeoSot replied:
I will leaf a max height for safety reasons. ok?

GeoSot commented:
@MartijnCuppens  I disagree with the "Provide some default theming (white background, maybe a border or shadow like the modals)" todo. 
I would like an agnostic component, and I am trying to save, some bytes, but as you insist, I added color, shadow etc

XhmikosR commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -1,7 +1,7 @@
 layout: docs
 title: Offcanvas
-description: Toggle the visibility sidebars across your project with a few classes and our JavaScript plugins.
+description: Toggle the visibility sidebars across your project with a few classes and our JavaScript plugin.
```
`visibility sidebars` doesn't seem right. Not sure how to reword this.

XhmikosR commented on site/content/docs/4.3/examples/offcanvas/index.html:
```diff
@@ -63,7 +60,7 @@
 <main role="main" class="container">
   <div class="d-flex align-items-center p-3 my-3 text-white-50 bg-purple rounded shadow-sm">
-    <img class="mr-3" src="/docs/{{< param docs_version >}}/assets/brand/bootstrap-outline.svg" alt="" width="48" height="48">
+    <img class="mr-3" src="/docs/{{< param docs_version >}}/assets/brand/bootstrap-outline.svg" alt="" width="48" height="48"/>
```
This is redundant, end tags are useless in HTML 5 :P

XhmikosR commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,200 @@
+</p>
+<div class="offcanvas offanvas-push-body bg-dark text-white" tabindex="-1" id="offcanvasExample">
+  <div class="offcanvas-header">
+     <div class="h4 col text-center">Offcanvas</div>
```
These examples need reindendation @GeoSot. All of them in this page.

It seems a `</div>` is missing somewhere in all of them.

MartijnCuppens commented:
> @MartijnCuppens
I disagree with the "Provide some default theming (white background, maybe a border or shadow like the modals)" todo.
I would like an agnostic component, and I am trying to save, some bytes, but as you insist, I added color, shadow etc

You definitely got a point about saving some bytes, but those variables can be set to `null` if people really want to generate as less CSS possible. We also got to be consistent with the other components (modals, cards, toasts), they have some basic styling applied to.

@twbs/css-review, what do you guys think?

GeoSot commented:
A new thought....... It could be responsive.  

Ex: .offcanvas-sm => hide sidebar bellow 'sm' breakpoint

XhmikosR commented:
@GeoSot please rebase and fix the linting errors (i.e don't use a section)

XhmikosR commented on site/content/docs/4.3/examples/offcanvas/index.html:
```diff
@@ -61,77 +58,91 @@
-      <small>Since 2011</small>
-    </div>
+<main role="main" class="container pt-5 d-flex flex-column align-items-center">
+  <div class="section">
```
Why is this needed? Just use our utils and drop the section divs

GeoSot committed:
Add a new offcanvas component

XhmikosR commented:
Alright, I had some time to do another quick review.

First of all, I want to say thanks for sticking with the PR. We are low on resources and we have so many things to do. I definitely want this added for v5, so I'll do my best to move forward with this.

1. you shouldn't have squashed the patches. It's not very ethical IMO. But what's done is done :)
2. the offcanvas example's mobile menu doesn't work anymore
3. the "blur" implementation isn't really a blur, but a `background-color`; we should be able to use `filter: blur(foo)`, /CC @MartijnCuppens 
4. when one clicks the right offcanvas and doesn't close it, and then clicks the bottom one, the bottom one shows up and the right one is automatically closed; I'm not sure if this is right, we need more people to chime in
5. we definitely need spell checking from a native English speaker (ping @mdo)
6. accessibility-wise, does this LGTY @patrickhlauke?

MartijnCuppens commented:
> the "blur" implementation isn't really a blur, but a background-color; we should be able to use filter: blur(foo), /CC @MartijnCuppens

The `filter` property can perform poorly when animated and doesn't work in IE. We might better keep the implementation but rename this to `backdrop`, just like the modals.

We should also keep the backdrop color the same as the modal for more consistency. Also the demos look a bit too colorful, just one example on how to change the background color should be enough. I also don't like the outlined focus state of the modal (Chrome), we should remove this (like we do with the modals).

XhmikosR commented:
Doesn't seem the blur is animated, though.

thehitechpanky commented:
Is this planned for v4 as well?

MartijnCuppens commented:
> Is this planned for v4 as well?

We're not planning on adding new features for v4 (bugfixes only)

GeoSot commented:
Ok Let's sum up.

I have updated the 'todos' on first message and I have to work the following


-    "the offcanvas example's mobile menu doesn't work anymore" 
It's true. in order to re-create the same example, the 'offcanvas' component has to be breakpoint -activated or to override it through css. Which approach should I follow?    **(need decision to continue)**

-  "blur" implementation has to be renamed to "backdrop" and try to keep a consistency with modal's implementation

-    we have to decide, the offcanvas behavior when another one opens **(need decision to continue)**

MartijnCuppens commented:
@GeoSot,

Please also address these issues I mentioned:
> We should also keep the backdrop color the same as the modal for more consistency. Also the demos look a bit too colorful, just one example on how to change the background color should be enough. I also don't like the outlined focus state of the modal (Chrome), we should remove this (like we do with the modals).

> we have to decide, the offcanvas behavior when another one opens (need decision to continue)

We don't support this with our modals. If we are going to support this, imo, the most logic approach is close the current offcanvas and open the next one. Supporting two open offcanvas elements will be an accessibility hell I guess.

GeoSot commented:
Hello guys. May I have some feedback to keep the request rolling?

@patrickhlauke can you help with accessibility (todo on first post)?
@XhmikosR  do I have to do something with the  FPS measurement?
@MartijnCuppens any comment after the last changes?

XhmikosR commented:
@GeoSot sorry for the radio silence.

Please bear with us, I'm trying to get everyone together so that we can move faster. Don't let this get you down, I hope we'll manage to regroup soon.

MartijnCuppens commented:
@GeoSot, Sure:

- The blue outline still feels weird (chrome):
![image](https://user-images.githubusercontent.com/11559216/73682394-98d54400-46c0-11ea-8f81-2eccbd0de3b9.png)
- The first example of the offcanvas can be opened with a `<a>`. I'm not sure that's a good idea accessibility-wise (ping @patrickhlauke)
- I don't think it's a good idea to use that many background colors in our examples. Maybe it's better to bundle colored examples under the `Color schemes
` section?
- There's still some _blur_ copy left in the docs

GeoSot commented:
@MartijnCuppens 

> The blue outline still feels weird (chrome):

I think the blue outline is because of  :focus styling (which is not mine). I have removed the outlines & shadows.  Do you want to override the default focus??

>The first example of the offcanvas can be opened with a <a>. I'm not sure that's a good idea accessibility-wise 

I tried to copy the other components behavior. So yes, if we want, I can remove it

>I don't think it's a good idea to use that many background colors in our examples. Maybe it's better to bundle colored examples under the Color schemes section?

I will handle it. I have changed only the example. my mistake ;)

MartijnCuppens commented:
> I think the blue outline is because of :focus styling (which is not mine). I have removed the outlines & shadows. Do you want to override the default focus??

Indeed, those are browser defaults. We also remove them for the modals, so let's get rid of it.

> > The first example of the offcanvas can be opened with a . I'm not sure that's a good idea accessibility-wise.

> I tried to copy the other components behavior. So yes, if we want, I can remove it

Leave it for now, unless @patrickhlauke has a strong opinion about this.

leomp12 commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+</table>
+
+{{< highlight js >}}
+var myOffcanvas = document.getElementById('myCollapsible')
```
```suggestion
var myOffcanvas = document.getElementById('myOffcanvas')
```

leomp12 commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+
+{{< highlight js >}}
+var myOffcanvas = document.getElementById('myCollapsible')
+myCollapsible.addEventListener('hidden.bs.offcanvas', function () {
```
```suggestion
myOffcanvas.addEventListener('hidden.bs.offcanvas', function () {
```

leomp12 reviewed (commented):
Hey @GeoSot ,
Just leaving minor fixes on last JS sample snippet (docs)

hassanbinsafdar commented:
if its possible to make off-canvas with body animate to right like reveal

mdo commented on scss/_offcanvas.scss:
```diff
@@ -0,0 +1,74 @@
+  color: $offcanvas-color;
+  background-color: $offcanvas-bg-color;
+  border: $offcanvas-border-width solid $offcanvas-border-color;
+  // Remove focus outline from opened offcanvas
```
```suggestion
```

mdo commented on site/assets/scss/docs.scss:
```diff
@@ -51,3 +51,8 @@
+
+.offcanvas{
+  z-index: 1072 // over .bd-subnavbar
+}
```
Move these styles to the appropriate section in `_component-examples.scss`, please. Also remember to add a space before the opening curly brace.

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+---
+layout: docs
+title: Offcanvas
+description: Toggle the visibility sidebars across your project with a few classes and our JavaScript plugin.
```
```suggestion
description: Build hidden sidebars into your project for navigation, shopping carts, and more with a few classes and our JavaScript plugin.
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+
+## How it works
+
+The offcanvas JavaScript plugin is used to show and hide a sidebar (or bottom-bar). Buttons or anchors are used as triggers, to specific elements you toggle.
```
```suggestion
The offcanvas JavaScript plugin shows and hides sidebar on the left, right, or bottom of your viewport. Buttons or anchors are used as triggers that are attached to specific elements you toggle.
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+
+The offcanvas JavaScript plugin is used to show and hide a sidebar (or bottom-bar). Buttons or anchors are used as triggers, to specific elements you toggle.
+
+Given how CSS handles animations, you cannot use `margin` & `translate` on a `.offcanvas` element. Use the class on an independent wrapping element.
```
```suggestion
Given how CSS handles animations, you cannot use `margin` or `translate` on a `.offcanvas` element. Instead, use the class as an independent wrapping element.
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+
+## Example
+
+Click the buttons below to show and hide another element via class changes:
```
```suggestion
Click the buttons below to show and hide an offcanvas element via class changes:
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+</p>
+<div class="offcanvas offanvas-push-body bg-dark text-white" tabindex="-1" id="offcanvasExample">
+  <div class="offcanvas-header">
+    <div class="h4 col text-center">Offcanvas</div>
```
```suggestion
    <div class="h4">Offcanvas</div>
```

mdo commented on scss/_variables.scss:
```diff
@@ -1224,6 +1225,25 @@ $close-color:                       $black !default;
+//Offcanvas
+
+$offcanvas-padding-y:               $modal-inner-padding !default;
+$offcanvas-padding-x:               $modal-inner-padding * 2 !default;
```
```suggestion
$offcanvas-padding-x:               $modal-inner-padding !default;
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+    Button with data-target
+  </button>
+</p>
+<div class="offcanvas offanvas-push-body bg-dark text-white" tabindex="-1" id="offcanvasExample">
```
`offanvas-push-body`?

GeoSot replied:
I bet it is left by mistake

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+      <span aria-hidden="true">&times;</span>
+    </button>
+  </div>
+  <div class="card card-body text-dark ">
```
I'd suggest nixing the bulk of this custom content and replacing it with something more functional and expected:

- Navigation elements (perhaps a stacked pill nav or a list group?)
- Simple `<p>`s of content

GeoSot replied:
@mdo 
I think now it is ok. Please take a look

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+{{< /example >}}
+
+## Position Modifiers
+
```
Add an intro line:

> Change the placement of an offcanvas element with modifier classes:

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+</div>
+{{< /example >}}
+
+## Position Modifiers
```
We use sentence case in Bootstrap, so this would be `Position modifiers`, but I think we can leave just as `Position` for brevity's sake.

```suggestion
## Position
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+
+## Position Modifiers
+
+- `.offcanvas-right` modifier, places offcanvas on the right side
```
```suggestion
- `.offcanvas-right` places offcanvas on the right of the viewport
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+## Position Modifiers
+
+- `.offcanvas-right` modifier, places offcanvas on the right side
+- `.offcanvas-bottom` modifier, places offcanvas on the bottom of the screen
```
```suggestion
- `.offcanvas-bottom` places offcanvas on the bottom of the viewport
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+
+{{< example >}}
+<p>
+  <button class="btn btn-primary" type="button" data-toggle="offcanvas" data-target="#offcanvasExample2" aria-expanded="false" aria-controls="offcanvasExample2">Toggle Right offcanvas</button>
```
```suggestion
  <button class="btn btn-primary" type="button" data-toggle="offcanvas" data-target="#offcanvasExample2" aria-expanded="false" aria-controls="offcanvasExample2">Toggle right offcanvas</button>
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+{{< example >}}
+<p>
+  <button class="btn btn-primary" type="button" data-toggle="offcanvas" data-target="#offcanvasExample2" aria-expanded="false" aria-controls="offcanvasExample2">Toggle Right offcanvas</button>
+  <button class="btn btn-primary" type="button" data-toggle="offcanvas" data-target="#offcanvasExample3" aria-expanded="false" aria-controls="offcanvasExample3">Toggle Bottom offcanvas</button>
```
```suggestion
  <button class="btn btn-primary" type="button" data-toggle="offcanvas" data-target="#offcanvasExample3" aria-expanded="false" aria-controls="offcanvasExample3">Toggle bottom offcanvas</button>
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+</p>
+<div class="offcanvas bg-dark text-white offcanvas-right" tabindex="-1" id="offcanvasExample2">
+  <div class="offcanvas-header">
+    <div class="h4 col text-center">Offcanvas Right Side</div>
```
```suggestion
    <div class="h4">Offcanvas right</div>
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+</div>
+<div class="offcanvas bg-dark text-white offcanvas-bottom" tabindex="-1" id="offcanvasExample3">
+  <div class="offcanvas-header">
+    <div class="h4 col text-center">Offcanvas Bottom</div>
```
```suggestion
    <div class="h4">Offcanvas bottom</div>
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+
+## Color schemes
+
+Theming offcanvas is very easy by using our [color utilities]({{< docsref "/utilities/colors" >}}).
```
```suggestion
Easily style an offcanvas element with a different `background-color` or `color` with our [color utilities]({{< docsref "/utilities/colors" >}}).
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+## Color schemes
+
+Theming offcanvas is very easy by using our [color utilities]({{< docsref "/utilities/colors" >}}).
+Any combination of `bg-*` and `text-{color}` utilities, can give a different result
```
Safe to delete this, this is self explanatory.

```suggestion
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+
+{{< example >}}
+<p>
+  <button class="btn btn-primary" type="button" data-toggle="offcanvas" data-target="#offcanvasExample31" aria-expanded="false" aria-controls="offcanvasExample31">Colored Offcanvas</button>
```
```suggestion
  <button class="btn btn-primary" type="button" data-toggle="offcanvas" data-target="#offcanvasColored" aria-expanded="false" aria-controls="offcanvasColored">Colored offcanvas</button>
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+<p>
+  <button class="btn btn-primary" type="button" data-toggle="offcanvas" data-target="#offcanvasExample31" aria-expanded="false" aria-controls="offcanvasExample31">Colored Offcanvas</button>
+</p>
+<div class="offcanvas bg-primary text-white" data-body="scroll" tabindex="-1" id="offcanvasExample31">
```
```suggestion
<div class="offcanvas bg-primary text-white" data-body="scroll" tabindex="-1" id="offcanvasColored">
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+</p>
+<div class="offcanvas bg-primary text-white" data-body="scroll" tabindex="-1" id="offcanvasExample31">
+  <div class="offcanvas-header">
+    <div class="h4 col text-center">Hey I got New Colors</div>
```
```suggestion
    <div class="h4">Colored offcanvas</div>
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+
+{{< /example >}}
+
+## Body Options
```
```suggestion
## Options
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+
+## Body Options
+
+By default, when an offcanvas is visible, body cannot be scrolled. You can use the following data-options:
```
```suggestion
By default, when an offcanvas is visible, the `<body>` of your page cannot be scrolled. You can use the following data-options to change this behavior:
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+
+By default, when an offcanvas is visible, body cannot be scrolled. You can use the following data-options:
+
+- `data-body="scroll"` enables scrollBar on body when offcanvas is open
```
```suggestion
- `data-body="scroll"` enables scrolling on the `<body>` when offcanvas is open
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+By default, when an offcanvas is visible, body cannot be scrolled. You can use the following data-options:
+
+- `data-body="scroll"` enables scrollBar on body when offcanvas is open
+- `data-body="backdrop"` hides scrollBar and creates a backdrop on body, when offcanvas is open
```
```suggestion
- `data-body="backdrop"` disables scrolling and creates a backdrop over the `<body>` when offcanvas is open
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+</p>
+<div class="offcanvas bg-dark text-white" data-body="scroll" tabindex="-1" id="offcanvasExample4">
+  <div class="offcanvas-header">
+    <div class="h4 col text-center">Wow, body scroller is with us again</div>
```
```suggestion
    <div class="h4">Offcanvas with scrolling</div>
    <p>Try scrolling the rest of the page to see this option in action.</p>
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+</div>
+<div class="offcanvas bg-dark text-white" data-body="backdrop" tabindex="-1" id="offcanvasExample5">
+  <div class="offcanvas-header">
+    <div class="h4 col text-center">Body is covered by a backdrop</div>
```
```suggestion
    <div class="h4">Offcanvas with backdrop</div>
    <p>Try clicking on the page's content to toggle the offcanvas.</p>
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+
+## Accessibility
+
+Make sure to add `aria-expanded` to the control element. This attribute explicitly conveys the current state of the collapsible element tied to the control to screen readers and similar assistive technologies. If the collapsible element is closed by default, the attribute on the control element should have a value of `aria-expanded="false"`. If you've set the collapsible element to be open by def ...
```
Do we need to update the language here to be changed from `collapsible` to `offcanvas`? Or, to avoid duplication, should we point folks to this section in the collapse docs?

GeoSot replied:
I chose the 'link' way, to avoid extra text.

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+
+## Usage
+
+The offcanvas plugin utilizes a few classes to handle the heavy lifting:
```
```suggestion
The offcanvas plugin utilizes a few classes and attributes to handle the heavy lifting:
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+- `.offcanvas.show` shows the content
+- `.offcanvas-right` hides the offcanvas on the right
+- `.offcanvas-bottom` hides the offcanvas on the bottom
+- `data-body="scroll"` enables body scrollbar when offcanvas is open
```
```suggestion
- `data-body="scroll"` enables `<body>` scrolling when offcanvas is open
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+- `.offcanvas-right` hides the offcanvas on the right
+- `.offcanvas-bottom` hides the offcanvas on the bottom
+- `data-body="scroll"` enables body scrollbar when offcanvas is open
+- `data-body="backdrop"` hides scrollBar and add backdrop to body, when offcanvas is open
```
```suggestion
- `data-body="backdrop"` disables `<body>` scrolling and adds backdrop when offcanvas is open
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+- `.offcanvas-bottom` hides the offcanvas on the bottom
+- `data-body="scroll"` enables body scrollbar when offcanvas is open
+- `data-body="backdrop"` hides scrollBar and add backdrop to body, when offcanvas is open
+- Add a dismiss button with the `data-dismiss="offcanvas"` attribute, which triggers the JavaScript functionality. Be sure to use the `<button>` element with it for proper behavior across all devices.
```
Separate this out from the list of classes/attributes:

```suggestion

Add a dismiss button with the `data-dismiss="offcanvas"` attribute, which triggers the JavaScript functionality. Be sure to use the `<button>` element with it for proper behavior across all devices.
```

mdo commented on site/content/docs/4.3/components/offcanvas.md:
```diff
@@ -0,0 +1,217 @@
+
+### Via data attributes
+
+Just add `data-toggle="offcanvas"` and a `data-target` or `href` to the element to automatically assign control of one offcanvas element. The `data-target` attribute accepts a CSS selector to apply the offcanvas to. Be sure to add the class `offcanvas` to the offcanvas element. If you'd like it to default open, add the additional class `show`.
```
```suggestion
Add `data-toggle="offcanvas"` and a `data-target` or `href` to the element to automatically assign control of one offcanvas element. The `data-target` attribute accepts a CSS selector to apply the offcanvas to. Be sure to add the class `offcanvas` to the offcanvas element. If you'd like it to default open, add the additional class `show`.
```

mdo commented on site/content/docs/4.3/examples/offcanvas/index.html:
```diff
@@ -2,20 +2,20 @@
-  - src: "offcanvas.js"
-body_class: "bg-light"
+- "offcanvas.css"
 ---
```
This entire file needs to be rethought. There's no reason to remove the custom page content from this. I imagine this should instead be updated to leave the page content as-is, and change the behavior of the navigation sidebar.

GeoSot replied:
@mdo 

Sorry, I'm not sure I understood what you propose

XhmikosR commented on site/assets/scss/docs.scss:
```diff
@@ -51,3 +51,9 @@
+// Temp
+
+.offcanvas{
+  z-index: 1072 // over .bd-subnavbar
```
We should probably move this in the docs variables and reuse it.

GeoSot replied:
Or as override on `site/assets/scss/docs.scss` before  the general variables import, but css-linter is complaining.

XhmikosR replied:
@MartijnCuppens can you please have a look when you have some time?

I also added another TODO regarding the `h4`s

MartijnCuppens replied:
Why is this added? The z-index of the off canvas component is higher than `.bd-subnavbar` anyway?

GeoSot replied:
it was added because  in a previous version, .bd-subnavbar had 1071 z-index and was covering offcanvas. I can make a check to find if it is changed. 

(offcanvas is on hold about 10 months, so many things have changed and sure  I have missed some of them)

GeoSot replied:
Finally, it was redundant

XhmikosR committed:
offcanvas.js: switch to string constants and `event.key`

XhmikosR commented:
@GeoSot please make sure you do **not** overwrite my rebase.

@MartijnCuppens and I will have a look at the remaining things and let you know after that.

XhmikosR committed:
Remove unneeded code

MartijnCuppens committed:
Sass optimizations

MartijnCuppens commented:
Pushed some changes. What bothers me, is the inconsistency between this offcanvas component and the modals (for example the way the backdrop is set). We got to make sure to keep this consistent (maybe we should make a backdrop component?).

GeoSot commented:
@MartijnCuppens during the last commit you removed the choice to change size (width or height) below a breakpoint (ex:mobile devices). Are you sure you want to remove it?

MartijnCuppens commented:
> @MartijnCuppens during the last commit you removed the choice to change size (width or height) below a breakpoint (ex:mobile devices). Are you sure you want to remove it?

Yes, we try to avoid `mobile` variables. Smaller viewports != mobile. Widths can be changed with the Sass variables.

XhmikosR commented:
Rebased. We should look into wrapping this up for v5.0.0-alpha1.

GeoSot committed:
Fixes

Make sure the element is hidden and not offscreen when inactive
fix close icon negative margins
Add content in right & bottom examples
Re-fix bottom offcanvas height not to cover all viewport

719media commented:
One thing that's always annoyed me with body overflow: hidden is with scroll bar width (e.g. Edge on windows, or if you have Show Scroll bars: always on your mac OS settings).

<img width="645" alt="Screen Shot 2020-06-19 at 1 33 57 PM" src="https://user-images.githubusercontent.com/5942229/85177922-0332ef00-b232-11ea-99be-588d749c50d6.png">


You get a jutter when applying overflow: hidden, just as the linked example for this PR. Anyway, I get around this by doing

```
const scrollBarGap = window.innerWidth - document.documentElement.clientWidth;
if (scrollBarGap > 0) {
  document.documentElement.style.setProperty('--scroll-bar-width', `${scrollBarGap}px`);
} else {
  document.documentElement.style.removeProperty('--scroll-bar-width');
}
```
In my javascript before applying the class that applies body overflow: hidden. Then, in that same class, I use the css
```
padding-right: var(--scroll-bar-width, 0);
```

Anyway, just a thought in case you find it useful. I'm not sure how the bootstrap project considers references to `window`, so there may be a different opinion there.

dotnetshadow commented:
Would this component also handle this style of offcanvas, where the sidebar would push the body content as well? https://www.w3schools.com/howto/howto_js_off-canvas.asp

GeoSot commented:
> 
> 
> Would this component also handle this style of offcanvas, where the sidebar would push the body content as well? https://www.w3schools.com/howto/howto_js_off-canvas.asp

I am afraid not with this implementation. Maybe on a future update

GeoSot commented:
@719media 

Please make a reminder again, when you see this 'pr' is going to be merged (or after it). It is too early to add more things and as you see it is already difficult to proceed on this process

vosidiy commented:
Actually it is easy to convert modal component into offcanvas. 
Here is I made https://bootstrap-menu.com/demos/modal.html 
Why not add special class like .modal-aside-right  and  .modal-aside-left  to make offcanvas from modal itself .

GeoSot commented:
> Actually it is easy to convert modal component into offcanvas.
> Here is I made https://bootstrap-menu.com/demos/modal.html
> Why not add special class like .modal-aside-right and .modal-aside-left to make offcanvas from modal itself .

I  had in mind a class that extends modal, but you went much further.
Yes personally i like it. 
It's a bit confusing as concept,  but it is simple with some lines of pure css gaining smaller size on final result and already tested js code

@mdo 
@XhmikosR 
@MartijnCuppens 


Please take a look on this.
It could be a good idea

ux-engineer commented:
Please consider supporting these kind of features: https://github.com/bootstrap-vue/bootstrap-vue/issues/5095

Just finished a custom implementation of a similar sidebar main navigation (also using collapsable submenus). Implemented it first without, and now with, BSV Sidebar component. Turned out to be much more work than anticipated. And still need to optimize for narrow mobile widths...

Should be pretty common layout for a modern web app that is optimized for mobile use.

mdo commented:
Capturing some of these latest comments in a new PR: see #31285. I know there have been previous efforts on this as well, and I appreciate @vosidiy for sharing their own here again. Please head over there to share any feedback if y'all have any!

vahidalvandi commented:
please add rtl support

happynewlife commented:
Thank you very much for adding rtl support.
I compared:
https://deploy-preview-30980--twbs-bootstrap.netlify.app/docs/5.0/cheatsheet/
with:
https://deploy-preview-30980--twbs-bootstrap.netlify.app/docs/5.0/cheatsheet-rtl/

I have to say it's more than very good, It's one of the best. 

Here are some thoughts:
1. Typography: every thing look fine. Thank you for that.
2. Images: Totally fine. nice.
3. Tables: Lovely.
4. Forms: Everything is nice.
5. Input group:
FYI: Email, URL and numbers in RTL languguages are LTR. Must of the time username is English (LTR).
![image](https://user-images.githubusercontent.com/65445744/88561893-40ff1080-d045-11ea-9c9a-0f82c76b6808.png)

6. Validation: great. very great. very very very great. Thanks. It's save a lots of work. Beautiful.
7. Alerts: Lovely.
8. Breadcrumb: seems nice to me.
9. Buttons: good
10. Dropdowns: By default it most starts from right to left:
![image](https://user-images.githubusercontent.com/65445744/88559343-f334d900-d041-11ea-884a-4c1a81e1a8db.png)
and also:
![image](https://user-images.githubusercontent.com/65445744/88559731-72c2a800-d042-11ea-96d0-b584f0a6caff.png)
and:
![image](https://user-images.githubusercontent.com/65445744/88559967-bddcbb00-d042-11ea-97e5-76fc7928fb75.png)

11. list groups: nice.
12. Modal: Lovely.
13. Navs: Good.
14. Navbar: Very good.
15. Pagination: Nice work.
16. Popovers: Default is better to be on the left.
![image](https://user-images.githubusercontent.com/65445744/88560752-c255a380-d043-11ea-9e2d-7a4c8606ffa7.png)

17. Progress: Very nice, also ltr progress is also usable in rtl languages.
18. Scrollspy: Very good. 
19. Spinners: Thanks but why? (Clockwise in rtl countries is same as ltr countries). Really thanks for that honest work.
20. Toasts: I love you for that.
21. Tooltips: Beautiful.

Thanks again.
By the way my first language is Persian and it's RTL.

ffoodd commented:
@happynewlife there's a dedicated PR for RTL (#30980), this is off-topic on this one. Thanks for your feedback though, this is well aligned with what is being done.

m-peterman commented:
Another thing to consider is the type of effect on off canvas. e.g. Switching between overlay and push(you could even consider putting in push where is shrinks the main content or 1 where is pushes this content off the other side off the canvas if off canvas expanded.)
see below for an example of what i mean.
https://tympanus.net/codrops/2013/04/17/slide-and-push-menus/

ux-engineer commented:
@m-peterman also I'd suggest to consider that kind of use case, that the 'offcanvas' sidebar would not be entirely hidden but just minimized. For example showing only icons while in expanded state having also link texts. That kind of layout seems pretty common need when implementing navigations.

HansUXdev commented:
Hey, if anyone wants help with the documentation on this tag me.
I'm maintaining a bootstrap 5 extension that already uses the off-canvas example template so I want to add this to it the second it gets merged into the main branches.

mdo commented:
Thinking we'll still be going with side modals instead of a brand new component. Please see #31285 and weigh in with any feedback or contributions via PR if y'all have ideas!

HansUXdev commented:
@mdo yeah, I saw that and I'm keeping an eye on it.
Should this issue be closed then?

GeoSot commented:
@mdo may I propose something different? 

I agree that till now both components have the same functionality. So yes it' a good idea the side modal. 
The disadvantages of the 'side' implementation, are that both components are chained by the same css and same functionality, and if you choose right now to keep them as one, they may be tied for ever.

I think it could be better to make an new component(js class), named as offcancvas, which for now will **extend** the **modal** js implementation and tests, so there is not need for extra writing, but with its **own css** classes.

Adapting this idea, will give the choice for a future decoupling or extra features, for both components

xcartmods commented:
https://github.com/xcartmods/bootstrap-4-side-modals

Boberro commented:
Hi, @GeoSot I found an issue with current version of a script. If a button (or link) used as a toggle to show an OffCanvas component has non-text elements, eg. `<svg>` or `<img>` tags, button only works if you click on the padding area of the button/link.

https://codepen.io/demicanadian/pen/LYZErQP

GeoSot commented:
@Boberro I confirm the case. 
For now I do not want spend more time  to make a fix. 
I do not have any new feedback to proceed on this PR, and as I see side-modal is in favor to be the new offcanvas

XhmikosR committed:
Wording tweaks

GeoSot committed:
update

GeoSot committed:
update tests and offcanvas class

GeoSot committed:
separate scrollbar functionality and use it in offcanvas

GeoSot committed:
Update .bundlewatch.config.json

patrickhlauke commented:
checked https://deploy-preview-29017--twbs-bootstrap.netlify.app/docs/5.0/components/offcanvas/ and it has major focus / focus order issues.

navigating through the page just with `tab`/`shift`+`tab`, the offcanvas can already be navigated into even when not visible (resulting in focus getting lost into the currently invisible offcanvas).

when the offcanvas is triggered, focus is not automatically moved to the offcanvas. in fact, it seems to reset itself to the start of the document (so on next `tab` you're all of a sudden on the "skip to content" link right at the top of the page).

once the user IS in the offcanvas, they can still tab back out of it.

at a high level, offcanvas should behave like a modal. this includes the way it handles focus.

having `aria-expanded="true"` on the offcanvas container makes no sense.

`aria-offcanvas="true"` makes absolutely no sense, that attribute does not exist.

patrickhlauke commented on js/src/offcanvas.js:
```diff
@@ -0,0 +1,222 @@
+    this.hidden = false
+    setTimeout(() => {
+      this._element.setAttribute('aria-expanded', true)
+      this._element.setAttribute('aria-offcanvas', true)
```
this attribute does not exist...

GeoSot replied:
Thank you @patrickhlauke for your feedback 

Please when you have time, take a second look. I think I have fixed indicated problems

patrickhlauke reviewed (changes requested):
see previous comment for various other quite fundamental issues, beyond this code one

GeoSot committed:
fix focus

mariusa commented:
Will this make it to 5.0, please?

patrickhlauke commented:
Testing https://deploy-preview-29017--twbs-bootstrap.netlify.app/docs/5.0/components/offcanvas/

* once the offcanvas is closed, focus / tabbing is broken in the page; can't reach anything when using `Tab`/`Shift`+`Tab`
* offcanvas falls below the search bar in the docs ![image](https://user-images.githubusercontent.com/895831/103490598-df717800-4e14-11eb-8ac4-4cff88e91a26.png)
* the offcanvas is, in essence, a modal dialog that just happens to show on the side rather than centered. it should have all the same things as modal dialog, e.g. `role="dialog"`, `aria-modal="true"` etc
* once the offcanvas is closed, focus should be programmatically set back to the control that originally opened the offcanvas

GeoSot committed:
update btn-close / fix focus on close

GeoSot committed:
add aria-modal and role
return focus on trigger when offcanvas is closed
change body scrolling timings

GeoSot commented:
Thank you again @patrickhlauke 

The issue with search bar z-index it happens on modal too cause the search bar on the docs css has the biggest z-index of all.
So I didn't change something, in case you want to fix it on another scope
![Capture](https://user-images.githubusercontent.com/22406063/103715682-ed1c2e80-4fc9-11eb-8af1-715ca2ab0753.PNG)

newyhj commented:
hope 5.0 release publish this features

patrickhlauke commented:
@GeoSot thanks for making the changes requested - can confirm that this now behaves like a modal. One last suggestion: I'd make the title of the offcanvas an actual heading (in theory it should be `<h1>`, but probably something like an `<h2>` or `<h3>` would be acceptable too), give it an `id`, and reference it from the modal container using `aria-labelledby` (same as with the regular modals) - this gives the modal a label that is then announced by assistive technologies when the modal first pops up and focus is moved to it.

patrickhlauke commented:
regarding the subnavbar displaying over the offcanvas, i filed https://github.com/twbs/bootstrap/pull/32839

patrickhlauke commented:
oh, additionally, would be good to swap out the lorem ipsum text with some actual/pseudo text (see https://github.com/twbs/bootstrap/pull/32246 and a few related recent PRs)

GeoSot committed:
move common code to reusable functions

GeoSot committed:
add aria-labelledby

GeoSot committed:
Replace lorem ipsum text

patrickhlauke commented:
when the off-canvas is hidden, but has keyboard-focusable elements, it is still in the focus order (`aria-hidden="true"` hides its content from AT, but does not prevent focusable controls from being focused). when not visible, the off-canvas needs to be fully neutralised (e.g. with `display:none`), the same way modal dialogs are.

color schemes and options examples should also have an explicit "X" close button. otherwise, a touch+AT user will not be able to close the off-canvas (nothing to swipe to and then click/activate to close it, and no ability to press `Esc`).

accessibility section points to collapse component accessibility docs, but really this should say that functionally, an off-canvas is like a modal dialog, that just happens to be positioned to the side. and then i'd say copy (rather than reference) and adapt the accessibility guidance for modal https://getbootstrap.com/docs/5.0/components/modal/#accessibility

a more high-level question (@mdo et al): do we call it "offcanvas" (one word, ugly) or "off-canvas", or even "off-canvas panel"? i'd tend towards the latter.

some of the wording in the docs part may also need a bit of tweaking, but i can get around to that once we're closer to finished.

ux-engineer commented:
What about 'Sidebar Navigation/Panel' or 'Side Panel'?

GeoSot commented:
@patrickhlauke you got me in thoughts. Should support swipe to close too?

patrickhlauke commented:
> @patrickhlauke you got me in thoughts. Should support swipe to close too?

i'd stay away from that to be honest, as it's...non-trivial (and swipe support would still not work for touch+AT users, as swipe gestures are handled directly by AT to control it first and foremost, unless a user explicitly does a pass-through gesture)

newyhj commented:
@GeoSot  wait for you,everyone needs this feature.

thor456 commented:
Offcanvas is an essential feature, and required feature. 

In Dashboard and Admin themes, the only required thing is Offcanvas. 
The developers need to understand this and make this permanent feature of `Bootstrap 5`

GeoSot committed:
fix focus when offcanvas is closed

GeoSot committed:
updates

mdo commented:
@geosot I owe you a huge apology for the delays here. Thank you for continuing to update and contribute elsewhere. After talking with @XhmikosR and re-reviewing your PR here, I'd love to make this happen in v5 if you're still up for it. No more delays on our part, we'd love to get this in the core.

Here's what @XhmikosR and I discussed over Slack:

- [ ] Finish and merge this PR
- [ ] Close any other PRs
- [ ] Iterate on the docs a bit more
- [ ] Detach and de-dupe the backdrop in #32439

Anything else we'd want to do?

GeoSot commented:
@mdo  I would like to super thank you. It is almost an honor for me to achieve this level of contribution. I 'll try my best

newyhj commented:
@mdo so,v5 beta2 can publish this feature?

GeoSot committed:
revert modal, add tests for scrollbar

GeoSot committed:
show backdrop by default

thor456 commented:
Bootstrap Offcanvas should be visible in large screen, only hide in smaller screen. 

Example: w3schools, tailwind. 
Visible in large screen, collapse with hamburger in small(540px) screen. 

For the intended behavior, take as example [codeply-offcanvas](https://www.codeply.com/p/FpImnZeaPt), 

this is how I am looking for the **Offcanvas**. I only need official support, and official touch - so that this does not get broken in say other browsers, mobile screens, other devices from other geographies. 

Uses: 
1. Tutorial Websites like w3schools
2. Dashboard theme(most used purpose)

patrickhlauke commented:
> a more high-level question (@mdo et al): do we call it "offcanvas" (one word, ugly) or "off-canvas", or even "off-canvas panel"? i'd tend towards the latter.

Still angling for at least the first hyphenated naming - @mdo thoughts?

vanillajonathan commented:
The classes are currently named `.offcanvas`, `.offcanvas-right` and `.offcanvas-bottom`. Considering the recent move to RTL, should they be named that, or should they be named `.offcanvas-start` and `.offcanvas-end` instead?

XhmikosR commented on site/content/docs/5.0/examples/offcanvas/offcanvas.css:
```diff
@@ -1,67 +1,3 @@
-  overflow-x: hidden; /* Prevent scroll on narrow devices */
-}
-
 body {
```
Would be nice if we could replace this with a core util and get rid of the CSS file here. Maybe `pt-5` is enough @mdo?

patrickhlauke commented:
Thank you for tackling the things I mentioned in https://github.com/twbs/bootstrap/pull/29017#issuecomment-766848961

Last few comments:

* remove the `aria-expanded="false"` from the various trigger elements. it's incorrect to use these (it's not an open/close disclosure widget here)
* wondering if for the code samples it would make sense to shorten the actual shown markup / only show `...` for the actual content of the panel? the samples appear super long and complex, but most the complexity is just all the example markup stuff in the off-canvas itself. i'd shorten them samples to basically show what's needed for the trigger and what the most important constituent parts of the panel are, e.g.

```
<div class="offcanvas offcanvas-left" tabindex="-1" id="offcanvasExample" aria-labelledby="exampleOffCanvasLiveLabel">
  <div class="offcanvas-header">
    <h5 class="modal-title" id="exampleOffCanvasLiveLabel">offcanvas</h5>
    <button type="button" class="btn-close text-reset" data-bs-dismiss="offcanvas" aria-label="Close">
    </button>
  </div>
  <div class="offcanvas-body">
    ...
  </div>
</div>
```

* I would still love to see (at least in the documentation) a better name for this. "Offcanvas" as one word is awkward. still angling for "Off-canvas" (which seems quite popular as a term, from some googling). the class names can remain `.offcanvas`, I don't mind that, but in documentation, I'd strongly favour the hyphenated version. 
* There's some tweaks to the wording in the documentation that I'd do, but I'll go for that in a separate PR once this is merged.

patrickhlauke commented on site/content/docs/5.0/components/offcanvas.md:
```diff
@@ -0,0 +1,259 @@
+
+## How it works
+
+The offcanvas JavaScript plugin shows and hides sidebar on the left, right, or bottom of your viewport. Buttons or anchors are used as triggers that are attached to specific elements you toggle.
```
"shows and hides sidebar on the left..." > "shows and hides an off-canvas panel on the left..." (or something similar)

patrickhlauke commented on site/content/docs/5.0/components/offcanvas.md:
```diff
@@ -0,0 +1,259 @@
+---
+layout: docs
+title: Offcanvas
```
Offcanvas > Off-canvas

(and throughout the docs page)

patrickhlauke commented on site/content/docs/5.0/components/offcanvas.md:
```diff
@@ -0,0 +1,259 @@
+You can use a link with the `href` attribute, or a button with the `data-bs-target` attribute. In both cases, the `data-bs-toggle="offcanvas"` is required.
+
+{{< example >}}
+<a class="btn btn-primary" data-bs-toggle="offcanvas" href="#offcanvasExample" role="button" aria-expanded="false" aria-controls="offcanvasExample">
```
remove the `aria-expanded="false"` here and in all other trigger elements. it's wrong.

patrickhlauke commented on site/content/docs/5.0/components/offcanvas.md:
```diff
@@ -0,0 +1,259 @@
+
+You can use a link with the `href` attribute, or a button with the `data-bs-target` attribute. In both cases, the `data-bs-toggle="offcanvas"` is required.
+
+{{< example >}}
```
for examples, which are extremely long and complex-looking mainly due to the amount of example content in the off-canvas, consider perhaps not using `{{< example >}}` and instead doing a split of actual code vs shown code, eliding the content of the `offcanvas-body` as it's not really that interesting from a sample code perspective.

see how https://github.com/twbs/bootstrap/blob/main/site/content/docs/5.0/components/modal.md for instance does it for vertically centered modals (https://getbootstrap.com/docs/5.0/components/modal/#vertically-centered)

patrickhlauke commented on site/content/docs/5.0/components/offcanvas.md:
```diff
@@ -0,0 +1,259 @@
+
+## Accessibility
+
+Be sure to add aria-labelledby="...", referencing the modal title, to .offcanvas. Note that you don’t need to add role="dialog" since we already add it via JavaScript.
```
I'd start this with a short mention a la "The off-canvas panel is, conceptually, a modal dialog."

vanillajonathan commented:
The classes are currently named `.offcanvas`, `.offcanvas-right` and `.offcanvas-bottom`. Considering the recent move to RTL, should they be named that, or should they be named `.offcanvas-start` and `.offcanvas-end` instead?

XhmikosR committed:
Update offcanvas.md

mdo committed:
Update offcanvas CSS to better match modals

- Add background-clip for borders
- Move from outline to border (less clever, more consistent)
- Add scss-docs in vars

mdo committed:
Revamp offcanvas docs

- Add static example to show and explain the components
- Split live examples and rename them
- Simplify example content
- Expand docs notes elsewhere
- Add sass docs

mdo committed:
Add .offcanvas-title instead of .modal-title

XhmikosR commented on site/content/docs/5.0/examples/offcanvas/index.html:
```diff
@@ -31,27 +28,27 @@
-            <li><a class="dropdown-item" href="#">Another action</a></li>
-            <li><a class="dropdown-item" href="#">Something else here</a></li>
-          </ul>
+          <div class="dropdown-menu" aria-labelledby="dropdown01">
```
Should we revert this change?

mdo committed:
Rename offcanvas example to offcanvas-navbar to reflect it's purpose

mdo committed:
labelledby references title and not header

mdo committed:
Add default shadow to offcanvas

GeoSot committed:
enable offcanvas-body to fill all the remaining wrapper area

GeoSot committed:
Be more descriptive, on Accessibility area

GeoSot committed:
remove redundant classes

GeoSot committed:
ensure in case of an already open offcanvas, not to open another one

GeoSot committed:
bring back backdrop|scroll combinations

XhmikosR committed:
Update offcanvas.md

GeoSot committed:
bring back toggling class

GeoSot committed:
refactor scrollbar method, plus tests

GeoSot commented:
Hey guys. I will need little help or your opinion.
Js tests for scrollbar fail in combination with modal's tests. I am very confident about the implementation and almost sure that it fails cause of some nailed bodystyle values on modal test. Both codes do their job (I think correct) but with different ways. Scrollbar.js uses only js and client widths, modal uses dummy html element. 
The difference on tests is that scrollbar.js tries to find out only if there is a scrollbar and its width only, as modal tests do some tricks with body right padding and though time are over-engineered on what this code can do.

The right way for me, is to fix modal tests but that's me. Is there another advice? Can anyone help to align them with reality? 

cc/
@mdo, @XhmikosR , @rohit2sharma95

GeoSot committed:
add check if element is not full-width, according to #30621

GeoSot committed:
revert all in modal

mdo commented:
I think I'd have to defer to @rohit2sharma95 and @XhmikosR on the best path forward there @geosot. The less clever a solution is though, the better it generally is. If we can streamline things and make it more accurate, that'd be the solution we should aim for IMO.

GeoSot committed:
use documentElement innerWidth

mdo committed:
Rename classes to -start and -end

Also copyedit some docs wording

mdo commented:
Pushed another update to edit some wording in the docs. For now, I'm leaving everything as `offcanvas`, from source code to usage in the docs copy. We can potentially change this later, but I prefer the one way to spell it, even though it's grammatically incorrect. 

I've also renamed the `-left` and `-right` classes to `-start` and `-end`, to match RTL naming.

From my POV, this is basically good to merge. Then I believe we can tackle the backdrop issue.

GeoSot committed:
omit some things on scrollbar

GeoSot committed:
PASS BrowserStack tests

-- IOS devices, Android devices and Browsers on Mac, hide scrollbar by default and appear it, only while scrolling.

rohit2sharma95 commented on js/src/offcanvas.js:
```diff
@@ -0,0 +1,239 @@
+const EVENT_KEY = `.${DATA_KEY}`
+const DATA_API_KEY = '.data-api'
+const ESCAPE_KEY = 'Escape'
+const DATA_BODY_ACTIONS = 'data-bs-body'
```
Better if you can accept this option in the configuration object (Like other plugins) :thinking:

GeoSot replied:
@Rohit. I really would prefer it to keep it simple right now, in order to proceed and to avoid more mess. 
If you find it so important, I can change it after. I am not very confident with the `Config` object, as every component uses it, in a different way

rohit2sharma95 commented on js/src/offcanvas.js:
```diff
@@ -0,0 +1,239 @@
+
+    this._isShown = element.classList.contains(CLASS_NAME_SHOW)
+    this._bodyOptions = element.getAttribute(DATA_BODY_ACTIONS) || ''
+    this._handleClosing()
```
Better if you name this like `_addEventListeners` :slightly_smiling_face: 
It keeps the consistent naming convention in the codebase

rohit2sharma95 commented on js/src/util/scrollbar.js:
```diff
@@ -0,0 +1,70 @@
+const SELECTOR_FIXED_CONTENT = '.fixed-top, .fixed-bottom, .is-fixed'
+const SELECTOR_STICKY_CONTENT = '.sticky-top'
+
+const getWidth = () => {
```
It would be good if you change its name so that it can be understood whose width you are getting :slightly_smiling_face:

GeoSot replied:
I am open to any proposal. I have started naming `getScrollbarWidth` etc but is seem too much. We are inside **scrollbar.js** and we need the width of scrollbar.

rohit2sharma95 commented on js/tests/unit/util/scrollbar.spec.js:
```diff
@@ -0,0 +1,182 @@
+    }
+  }
+
+  const scrollBarIsHidden = () => { // IOS devices, Android devices and Browsers on Mac, hide scrollbar by default and appear it, only while scrolling. So the tests for scrollbar would fail
```
It's better to prefix it with 'is' since it is like a boolean flag

rohit2sharma95 commented on js/src/offcanvas.js:
```diff
@@ -0,0 +1,239 @@
+  }
+
+  _bodyOptionsHas(option) {
+    return this._bodyOptions.split('|').includes(option)
```
Few plugins are already using comma separated values, so if you also use a comma here instead of a pipe, that would keep consistency IMO.

GeoSot replied:
I agree and I 've changed it. IMO I would prefer 'pipe'. Seems more programmatic in my eyes

GeoSot committed:
Rename '_handleClosing' to '_addEventListeners'

GeoSot committed:
change pipe usage to comma

XhmikosR committed:
Data.get

XhmikosR commented:
All right, this is pretty hard to rebase to clean up the patches into less commits, so I'll have to squash it into one patch.

Huge thanks to @GeoSot for sticking with us after all this time, to everyone else who contributed to the branch and to those who helped reviewing the PR!

We'll need to make some refactoring later to deduplicate some stuff, tweak doc examples, add a navbar offcanvas example, etc, but at least this should help a lot of people.

mdo commented:
Took us long enough 😆 ❤️

rohit2sharma95 commented:
Thank you George for the hard work :heart::blue_heart:  and for debugging BrowserStack tests :smile:

GeoSot commented:
Thank you all for your support on this MR. Thanks @XhmikosR  for your patience
